### PR TITLE
Load workspaces

### PR DESCRIPTION
**Pull request by maciektr**
**Stack**:
- #508
- #506
- #502
- #500
- #490
- #489
- #465
- #446
- #447
- #445 ⬅

⚠️ *Part of a stack created by [spr](https://github.com/ejoffe/spr). Do not merge manually using the UI - doing so may have unexpected results.*

### CODE REVIEW

File: scarb/src/core/manifest/maybe_workspace.rs [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+/// [`MaybeWorkspace::resolve`] much nicer
+///
+/// Implementors should have a field `workspace` with the type of `bool`. It is used to ensure
+/// `workspace` is not `false` in a `Cargo.toml`
+pub trait WorkspaceInherit {
+    /// This is the workspace table that is being inherited from.
+    /// For example `[workspace.dependencies]` would be the table "dependencies"
+    fn inherit_toml_table(&self) -> &str;
+
+    /// This is used to output the value of the implementors `workspace` field
```
Comment (mkaput) [inline scarb/src/core/manifest/maybe_workspace.rs:15]:
```suggestion
    /// This is used to output the value of the implementors `workspace` field.
```

File: scarb/src/core/manifest/maybe_workspace.rs [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+/// This Trait exists to make [`MaybeWorkspace::Workspace`] generic. It makes deserialization of
+/// [`MaybeWorkspace`] much easier, as well as making error messages for
+/// [`MaybeWorkspace::resolve`] much nicer
+///
+/// Implementors should have a field `workspace` with the type of `bool`. It is used to ensure
+/// `workspace` is not `false` in a `Cargo.toml`
+pub trait WorkspaceInherit {
+    /// This is the workspace table that is being inherited from.
+    /// For example `[workspace.dependencies]` would be the table "dependencies"
```
Comment (mkaput) [inline scarb/src/core/manifest/maybe_workspace.rs:12]:
```suggestion
    /// For example `[workspace.dependencies]` would be the table "dependencies".
```

**Review by mkaput — CHANGES_REQUESTED** [review]

File: scarb/src/core/workspace.rs [hunk 3]
```diff
@@ -1,31 +1,52 @@
...
-/// **Note:** Currently only single-package workspaces are supported.
-///
 /// A workspace is often created very early on and then threaded through all other functions.
 /// It's typically through this object that the current package is loaded and/or learned about.
 pub struct Workspace<'c> {
     config: &'c Config,
-    package: Package,
+    members: BTreeMap<PackageId, Package>,
+    root_package: Option<PackageId>,
+    manifest_path: Utf8PathBuf,
```
Comment (mkaput) [inline scarb/src/core/workspace.rs:24]: shouldn't we store manifest here?

File: scarb/src/core/workspace.rs [hunk 4]
```diff
@@ -52,12 +75,30 @@ impl<'c> Workspace<'c> {
     /// In this case an error is returned indicating that the operation
     /// must be performed on specific package.
     pub fn current_package(&self) -> Result<&Package> {
-        Ok(&self.package)
+        Ok(self.members.values().next().unwrap())
+    }
+
+    pub fn root_package(&self) -> Option<Package> {
+        self.root_package.and_then(|id| self.package(&id)).cloned()
+    }
```
Comment (mkaput) [inline scarb/src/core/workspace.rs:83]: what does this represent? the name is not self-descriptive unfortunately

File: scarb/src/core/workspace.rs [hunk 5]
```diff
@@ -52,12 +75,30 @@ impl<'c> Workspace<'c> {
...
+
+    pub fn package(&self, id: &PackageId) -> Option<&Package> {
+        self.members.get(id)
+    }
+
+    pub fn fetch_package(&self, id: &PackageId) -> Result<&Package> {
+        self.package(id)
+            .ok_or_else(|| anyhow::anyhow!("package `{}` is not a member of this workspace", id))
```
Comment (mkaput) [inline scarb/src/core/workspace.rs:91]:
```suggestion
            .ok_or_else(|| anyhow!("package `{}` is not a member of this workspace", id))
```

File: scarb/src/core/workspace.rs [hunk 6]
```diff
@@ -1,31 +1,52 @@
...
-    package: Package,
+    members: BTreeMap<PackageId, Package>,
+    root_package: Option<PackageId>,
+    manifest_path: Utf8PathBuf,
 }
 
 impl<'c> Workspace<'c> {
+    pub(crate) fn new(
+        manifest_path: Utf8PathBuf,
+        packages: Vec<Package>,
```
Comment (mkaput) [inline scarb/src/core/workspace.rs:30]: Please document that the first package will be the "current" one

Comment (mkaput) [inline scarb/src/core/workspace.rs:30]:
```suggestion
        packages: &[Package],
```

File: scarb/src/core/workspace.rs [hunk 7]
```diff
@@ -1,31 +1,52 @@
...
+        root_package: Option<PackageId>,
+        config: &'c Config,
+    ) -> Result<Self> {
+        let packages = packages
+            .iter()
+            .map(|p| (p.id, p.clone()))
+            .collect::<Vec<_>>();
```
Comment (mkaput) [inline scarb/src/core/workspace.rs:37]:
```suggestion
            .collect::<BTreeMap<_, _>>();
```

**Review by mkaput — CHANGES_REQUESTED** [review]

File: scarb/src/core/workspace.rs (hunk 4 above)
Comment (mkaput) [reply scarb/src/core/workspace.rs:83]: Oh I didn't know that feature. Yeah, the comment is enough

**Review by mkaput — CHANGES_REQUESTED** [review]

**Review by mkaput — CHANGES_REQUESTED** [review]

File: scarb/src/ops/workspace.rs [hunk 8]
```diff
@@ -29,23 +34,140 @@ pub fn read_workspace_with_source_id<'c>(
...
+                        .map(PathBuf::try_into_utf8)
+                })
+                .collect::<Result<Result<Vec<_>, _>>>()?
+        })
+        .collect::<Result<Vec<_>>>()
+        .map(|v| {
+            v.into_iter()
+                .flatten()
+                // Make sure all files exist.
+                .filter(|p| p.is_file())
```
Comment (mkaput) [inline scarb/src/ops/workspace.rs:164]: I wonder if we shouldn't print a warning if we matched a directory that is missing Scarb.toml

Technically, I think we should make users avoid overusing globs in this way? What do you think?

**Review by mkaput — APPROVED** [review]

File: scarb/src/core/workspace.rs (hunk 3 above)
Comment (mkaput) [reply scarb/src/core/workspace.rs:24]: bump

File: scarb/src/ops/workspace.rs (hunk 8 above)
Comment (maciektr) [reply scarb/src/ops/workspace.rs:164]: Hmm, do you think a warning is enough, or we should just error?

Comment (maciektr) [reply scarb/src/ops/workspace.rs:164]: Afaik Cargo does emit errors for such cases, but they also have `workspace.exclude` for better control over the search.

Comment (mkaput) [reply scarb/src/ops/workspace.rs:164]: warning should be enough I think

workspace.exclude is something we can leave in our backlog :)

**Review by mkaput — APPROVED** [review]

File: scarb/src/ops/workspace.rs (hunk 8 above)
Comment (mkaput) [reply scarb/src/ops/workspace.rs:164]: this change can be done in separate PR, resolving this discussion

File: scarb/src/core/workspace.rs (hunk 3 above)
Comment (maciektr) [reply scarb/src/core/workspace.rs:24]: I feel like `core::manifest::Manifest` is really concerned with being the **package manifest**, defining targets, dependencies etc. We could create some `WorkspaceManifest` representation, although I do not think we need it **yet**. All the values defined for `.workspace = true` manifest inheritance can be merged when deserializing package manifests (like we do as of this pr), without loosing any functionality.

Comment (mkaput) [reply scarb/src/core/workspace.rs:24]: Makes sense

**Review by mkaput — APPROVED** [review]

**Comment by maciektr** [comment]
✓ Commit merged in pull request [#446](https://github.com/software-mansion/scarb/pull/446)

**Pull request closed without merging**